#  Práctica 2 Fundamentos de Aprendizaje Automático

## Apartado 1: Particionado

A continuación se va a realizar un análisis de las tres estrategias de particionado usadas en la práctica para los conjuntos "balloons.data" y "tic-tac-toe.data".

### Validación Simple

In [1]:
%%python3
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('balloons.data')
estrategia=EstrategiaParticionado.ValidacionSimple(0.25,3)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador, seed=None)
for particion in estrategia.listaPartic:
    print("Train: ",particion.indicesTrain)
    print("Test: ", particion.indicesTest)

Train:  [15, 7, 2, 11, 0]
Test:  [12, 1, 6, 3, 19, 14, 16, 18, 5, 8, 13, 17, 9, 10, 4]
Train:  [2, 4, 5, 16, 8]
Test:  [10, 9, 3, 19, 1, 11, 14, 0, 6, 17, 7, 18, 12, 13, 15]
Train:  [18, 12, 2, 17, 19]
Test:  [3, 7, 13, 16, 11, 5, 10, 1, 15, 9, 14, 6, 8, 4, 0]


En el ejemplo hemos leido los datos de "ballons.data", hemos creado la estrategia llamando al clasificador y luego a la validación del mismo.
Al llamar a la validación se crean las particiones acorde a los argumentos introducidos, es decir, en este caso hemos creado una particion que dedica un 25% de los datos para train y el restante para test de forma aleatoria en 3 particiones diferentes.
Como se puede comprobar en la salida las listas de train y test entre ellas son totalmente diferentes, por lo que las permutaciones se hacen correctamente para que no se repitan particiones. Además, también se puede comprobar que un 25% de los elementos está en Train (en este caso 5 de 20) y el restante en Test.

In [11]:
%%python3
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('tic-tac-toe.data')
estrategia=EstrategiaParticionado.ValidacionSimple(0.25,3)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador, seed=None)
for particion in estrategia.listaPartic:
    print(len(particion.indicesTrain))
    print("Train: ",particion.indicesTrain, "\n")
    print(len(particion.indicesTest))
    print("Test: ", particion.indicesTest, "\n")

239
Train:  [615, 800, 573, 717, 498, 523, 830, 728, 478, 410, 651, 443, 813, 103, 950, 619, 258, 20, 873, 64, 560, 935, 714, 911, 400, 94, 782, 375, 508, 161, 90, 372, 70, 816, 752, 301, 888, 666, 461, 789, 580, 259, 213, 776, 940, 178, 724, 660, 785, 859, 598, 204, 266, 163, 409, 120, 525, 101, 559, 61, 95, 89, 941, 200, 953, 228, 905, 924, 306, 381, 852, 701, 485, 764, 951, 801, 479, 314, 37, 528, 621, 581, 943, 766, 243, 456, 758, 85, 429, 280, 533, 553, 489, 382, 111, 31, 115, 321, 312, 739, 2, 957, 193, 835, 503, 638, 6, 298, 57, 753, 430, 568, 377, 122, 215, 467, 887, 956, 822, 640, 939, 347, 777, 700, 370, 15, 83, 683, 411, 710, 579, 534, 299, 874, 92, 80, 896, 332, 441, 121, 338, 687, 798, 912, 21, 743, 66, 308, 823, 52, 678, 791, 747, 809, 902, 511, 480, 97, 431, 639, 340, 548, 257, 633, 442, 194, 946, 36, 209, 154, 229, 680, 591, 917, 877, 671, 376, 427, 754, 844, 42, 838, 544, 913, 522, 390, 1, 260, 119, 307, 562, 681, 650, 506, 317, 181, 129, 389, 922, 577, 587, 797, 589, 

Ejecutamos el mismo programa ahora con el archivo de datos "tic-tac-toe.data", este es mucho más extenso en comparación a "balloons.data".
El fichero"tic-tac-toe.data" tiene 958 filas de las cuales se cojen un 25% para Train lo que nos deja 239.5, como hay una fila que no se puede partir en dos al redondear a 239 se pierde y el restante se va a Test.
Como sucedía en el ejemplo anterior, las particiones son diferentes entre ellas.

### Validación Cruzada

In [30]:
%%python3
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('balloons.data')
estrategia=EstrategiaParticionado.ValidacionCruzada(4)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador, seed=None)
for particion in estrategia.listaPartic:
    print("Train: ",particion.indicesTrain)
    print("Test: ", particion.indicesTest)

Train:  [19, 6, 13, 9, 11, 5, 0, 18, 14, 15, 17, 7, 10, 16, 12]
Test:  [4, 2, 1, 8, 3]
Train:  [4, 2, 1, 8, 3, 5, 0, 18, 14, 15, 17, 7, 10, 16, 12]
Test:  [19, 6, 13, 9, 11]
Train:  [4, 2, 1, 8, 3, 19, 6, 13, 9, 11, 17, 7, 10, 16, 12]
Test:  [5, 0, 18, 14, 15]
Train:  [4, 2, 1, 8, 3, 19, 6, 13, 9, 11, 5, 0, 18, 14, 15]
Test:  [17, 7, 10, 16, 12]


En el ejemplo anterior, ejecutamos el mismo programa que en los casos anteriores usando la validación cruzada con 4 particiones.
En la impresión se puede comprobar como las particiones de test no se repiten y es un 1/4 de los datos (5 ya que en hay 20 datos). También se puede apreciar que las particiones de train son 3/4 de los datos y que en cada partición falta la particion de test y va rotando.

In [32]:
%%python3
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('tic-tac-toe.data')
estrategia=EstrategiaParticionado.ValidacionCruzada(4)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador, seed=None)
for particion in estrategia.listaPartic:
    print(len(particion.indicesTrain))
    print("Train: ",particion.indicesTrain)
    print(len(particion.indicesTest))
    print("Test: ", particion.indicesTest)

719
Train:  [407, 734, 663, 843, 0, 808, 43, 85, 31, 666, 650, 531, 620, 835, 298, 628, 543, 680, 51, 464, 398, 637, 382, 138, 955, 838, 355, 369, 859, 466, 676, 15, 692, 905, 529, 802, 170, 56, 572, 277, 883, 643, 242, 694, 497, 325, 159, 919, 397, 944, 568, 691, 837, 686, 405, 644, 744, 907, 214, 860, 99, 467, 287, 50, 68, 315, 141, 186, 14, 180, 943, 629, 533, 950, 210, 832, 502, 129, 814, 312, 32, 651, 916, 205, 564, 16, 343, 743, 849, 526, 935, 795, 306, 384, 354, 460, 928, 218, 746, 418, 492, 395, 605, 430, 779, 156, 286, 437, 671, 449, 520, 804, 393, 168, 94, 293, 475, 164, 128, 368, 785, 102, 423, 864, 584, 326, 527, 332, 151, 212, 388, 83, 5, 373, 677, 197, 896, 957, 272, 499, 52, 365, 274, 309, 452, 207, 844, 904, 98, 626, 223, 690, 82, 226, 831, 92, 185, 763, 65, 282, 313, 513, 78, 927, 174, 895, 335, 361, 415, 530, 351, 707, 49, 438, 902, 570, 500, 167, 589, 279, 540, 590, 807, 323, 702, 322, 109, 111, 523, 234, 575, 909, 803, 722, 122, 696, 646, 622, 280, 547, 229, 416, 83

958 Como en el caso de validación simple usando el archivo "tic-tac-toe.data" hay 958 datos, al realizar 4 particiones la partición test tiene 239 en la particion test y 719 datos en la partición train, por lo que podemos concluir que las particiones se realizan correctamente y viendo el ejemplo de "balloons.data" podemos afirmar que la validación cruzada se realiza correctamente.

### Validación por Bootstrap

In [33]:
%%python3
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('balloons.data')
estrategia=EstrategiaParticionado.ValidacionBootstrap(3)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador, seed=None)
for particion in estrategia.listaPartic:
    print("Train: ",particion.indicesTrain)
    print("Test: ", particion.indicesTest)

Train:  [16  9 12  0  2 19  6 16  4  3  7 13 17 17 14  8 12  9 19  7]
Test:  [1, 5, 10, 11, 15, 18]
Train:  [15 14  1 16  8 15 10 18  1 16  8 11 13 16 16  7  1  7 13 19]
Test:  [0, 2, 3, 4, 5, 6, 9, 12, 17]
Train:  [16  5  2 12  2 11 18 17  8  5 18  1  3  5 14  8 14 17 19 16]
Test:  [0, 4, 6, 7, 9, 10, 13, 15]


Se puede observar en los resultados que las particiones train y test tienen un tamaño variable y los datos en ambas particiones se pueden repetir. No sigue tantas restricciones como las anteriores validaciones y por lo tanto no hay mucho más que añadir.

In [35]:
%%python3
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('tic-tac-toe.data')
estrategia=EstrategiaParticionado.ValidacionBootstrap(4)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador, seed=None)
for particion in estrategia.listaPartic:
    print(len(particion.indicesTrain))
    print("Train: ",particion.indicesTrain)
    print(len(particion.indicesTest))
    print("Test: ", particion.indicesTest)

958
Train:  [850 541 148  58 811 482 933 570 438 262 411 379 318 515  62 264 698 848
 285 756 836 247 595  20 292 899 699 634 292 248 169 604 204 708 636 466
 666 439 186 636 729 388 850 284 748 488 433 684 225 605 862 638 224 841
  80 574 594 217 708 558 918 732 841 895 615 898 565 225 423 698 436 276
 855 448 224 164 154 612 867 748 407 922 875  29 123 519 545 805 652 426
 575 389 211 138 648 776 726 318  67 248 401 534 494  46 653 825 658 924
 399 353 287 743 399 353 586 584 889 655 240 207 513 207  91 449  58 693
 264 554 119 489 321 744 779 791  18  58 954 652 563 530 341 911 541 178
 893 115 924 870 795 297 731 770 241 155 490 390 501 935 289 456 820 527
 379 475 563 910 235 292 322 317   8 948 464 208 896 824 705 754 360 126
 902 558 517 295 885 671  54 490 362 209 181 767 352 712 477 473  50 213
 755 494  66 845 881 840 881 692 146 542 859 682 200 644 626 458  35 702
 448 286   9  55 326  33 125 252 535 664 822 852 820 906  69 935 929 416
 794 824 810 687 613 776  53 131 712 42

En el ejemplo ejecutado es más difícil comprobar que se realiza correctamente, se puede apreciar como la lista de índices de train hay 958 índices y test tiene menos índices que la lista de train. En el ejemplo con el archivo "tic-tac-toe.data" es bastante complicado ver que se realiza correctamente, pero viendo la ejecución del archivo "balloons.data" se puede apreciar que se realiza correctamente.

## Apartado 2: Naive-Bayes

En este apartado se va a realizar un análisis de los resultados de la ejecución para el conjunto de datos "tic-tac-toe.data" y "german.data".

### Naive-Bayes sin aplicar la corrección de Laplace

In [62]:
from Datos import Datos
import EstrategiaParticionado
import Clasificador

dataset=Datos('tic-tac-toe.data')
estrategia= EstrategiaParticionado.ValidacionCruzada(4)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador,laplace=0, seed=None)
print("Probabilidad de error: ",errores)
media = np.mean(errores)
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
print("Promedio de error: ", media)
print("Desviación Típica: ", d_tipica)

ZeroDivisionError: division by zero

EXPLICACION

In [55]:
from Datos import Datos
import EstrategiaParticionado
import Clasificador
import numpy as np

dataset=Datos('german.data')
estrategia= EstrategiaParticionado.ValidacionCruzada(4)
clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador,laplace=0, seed=None)
print("Probabilidad de error: ",errores)
media = np.mean(errores)
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
print("Promedio de error: ", media)
print("Desviación Típica: ", d_tipica)

/usr/lib/python3/dist-packages/numpy/core/fromnumeric.py:2909: RuntimeWarning: Mean of empty slice.
  out=out, **kwargs)
/usr/lib/python3/dist-packages/numpy/core/fromnumeric.py:3146: RuntimeWarning: Degrees of freedom <= 0 for slice
  **kwargs)


ZeroDivisionError: division by zero

Explicacion

### Naive-Bayes aplicando la corrección de Laplace

In [ ]:
from Datos import Datos
import EstrategiaParticionado
import Clasificador
import numpy as np

dataset=Datos('tic-tac-toe.data')
estrategia= EstrategiaParticionado.ValidacionCruzada(4)

clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador,laplace=1, seed=None)
print("Probabilidad de error: ",errores)
media = np.mean(errores)
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
print("Promedio de error: ", media)
print("Desviación Típica: ", d_tipica)

EXPLICACION

In [ ]:
from Datos import Datos
import EstrategiaParticionado
import Clasificador
import numpy as np

dataset=Datos('german.data')
estrategia= EstrategiaParticionado.ValidacionCruzada(4)

clasificador=Clasificador.ClasificadorNaiveBayes()
errores=clasificador.validacion(estrategia,dataset,clasificador,laplace=1, seed=None)
print("Probabilidad de error: ",errores)
media = np.mean(errores)
d_tipica = np.std(np.transpose(np.array(errores)), axis=0)
print("Promedio de error: ", media)
print("Desviación Típica: ", d_tipica)

EXPLICACION

## Apartado 3: Scikit-Learn

En el siguiente apartado se va a realizar el mismo análisis que en el apartado 2 pero usando los métodos del paquete scikit-learn.

## Apartado 4: Evaluación de hipótesis mediante Análisis ROC